From 077e53ccc316091503fcd7cdc5c7b2f5ad5103dc Mon Sep 17 00:00:00 2001
From: Arthur Schiwon <blizzz@arthur-schiwon.de>
Date: Wed, 31 Jul 2019 00:53:20 +0200
Subject: [PATCH] instead of upsert, fallback to default query on PgSQL <= 9.4

because there is no upsert yet

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>
---
diff --git a/lib/private/DB/AdapterPgSql.php b/lib/private/DB/AdapterPgSql.php
index af1978d051c..42e57cd45f7 100644
--- a/lib/private/DB/AdapterPgSql.php
+++ b/lib/private/DB/AdapterPgSql.php
@@ -24,7 +24,11 @@
 
 namespace OC\DB;
 
+use Doctrine\DBAL\DBALException;
+
 class AdapterPgSql extends Adapter {
+	protected $compatModePre9_5 = null;
+
 	public function lastInsertId($table) {
 		return $this->conn->fetchColumn('SELECT lastval()');
 	}
@@ -40,12 +44,29 @@ public function fixupStatement($statement) {
 	 * @suppress SqlInjectionChecker
 	 */
 	public function insertIgnoreConflict(string $table,array $values) : int {
+		if($this->isPre9_5CompatMode() === true) {
+			return parent::insertIgnoreConflict($table, $values);
+		}
+
+		// "upsert" is only available since PgSQL 9.5, but the generic way
+		// would leave error logs in the DB.
 		$builder = $this->conn->getQueryBuilder();
 		$builder->insert($table);
-		foreach($values as $key => $value) {
+		foreach ($values as $key => $value) {
 			$builder->setValue($key, $builder->createNamedParameter($value));
 		}
 		$queryString = $builder->getSQL() . ' ON CONFLICT DO NOTHING';
 		return $this->conn->executeUpdate($queryString, $builder->getParameters(), $builder->getParameterTypes());
 	}
+
+	protected function isPre9_5CompatMode(): bool {
+		if($this->compatModePre9_5 !== null) {
+			return $this->compatModePre9_5;
+		}
+
+		$version = $this->conn->fetchColumn('SHOW SERVER_VERSION');
+		$this->compatModePre9_5 = version_compare($version, '9.5', '<');
+
+		return $this->compatModePre9_5;
+	}
 }
