diff --git a/core/Command/App/Install.php b/core/Command/App/Install.php
index 4e9c846cbd4b2..eab0a294cc367 100644
--- a/core/Command/App/Install.php
+++ b/core/Command/App/Install.php
@@ -9,6 +9,7 @@
 namespace OC\Core\Command\App;
 
 use OC\Installer;
+use OCP\App\AppPathNotFoundException;
 use OCP\App\IAppManager;
 use Symfony\Component\Console\Command\Command;
 use Symfony\Component\Console\Input\InputArgument;
@@ -58,9 +59,11 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 		$appId = $input->getArgument('app-id');
 		$forceEnable = (bool)$input->getOption('force');
 
-		if ($this->appManager->isInstalled($appId)) {
+		try {
+			$this->appManager->getAppPath($appId);
 			$output->writeln($appId . ' already installed');
 			return 1;
+		} catch (AppPathNotFoundException) {
 		}
 
 		try {
